pkg/middlewares/headers/headers.go

```diff
@@ -102,33 +108,121 @@ func (s secureHeader) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
-	customRequestHeaders map[string]string
+type Header struct {
+	next         http.Handler
+	headers      *config.Headers
+	originHeader string
 }
 
 // NewHeader constructs a new header instance from supplied frontend header struct.
-func newHeader(next http.Handler, headers config.Headers) *header {
-	return &header{
-		next:                 next,
-		customRequestHeaders: headers.CustomRequestHeaders,
+func NewHeader(next http.Handler, headers config.Headers) *Header {
+	return &Header{
+		next:    next,
+		headers: &headers,
 	}
 }
 
-func (s *header) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+func (s *Header) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	reqAcMethod := req.Header.Get("Access-Control-Request-Method")
+	reqAcHeaders := req.Header.Get("Access-Control-Request-Headers")
+	s.originHeader = req.Header.Get("Origin")
+
+	if reqAcMethod != "" && reqAcHeaders != "" && s.originHeader != "" && req.Method == http.MethodOptions {
+		// If the request is an OPTIONS request with an Access-Control-Request-Method header, and  Access-Control-Request-Headers headers,
+		// and Origin headers, then it is a CORS preflight request, and we need to build a custom response: https://www.w3.org/TR/cors/#preflight-request
+		if s.headers.AccessControlAllowCredentials {
+			rw.Header().Set("Access-Control-Allow-Credentials", "true")
+		}
+
+		allowHeaders := strings.Join(s.headers.AccessControlAllowHeaders, ",")
+		if allowHeaders != "" {
+			rw.Header().Set("Access-Control-Allow-Headers", allowHeaders)
+		}
+
+		allowMethods := strings.Join(s.headers.AccessControlAllowMethods, ",")
+		if allowMethods != "" {
+			rw.Header().Set("Access-Control-Allow-Methods", allowMethods)
+		}
+
+		allowOrigin := s.getAllowOrigin()
+
+		if allowOrigin != "" {
+			rw.Header().Set("Access-Control-Allow-Origin", allowOrigin)
+		}
+
+		rw.Header().Set("Access-Control-Max-Age", strconv.Itoa(int(s.headers.AccessControlMaxAge)))
+
+		return
+	}
+
 	s.modifyRequestHeaders(req)
-	s.next.ServeHTTP(rw, req)
+	// If there is a next, call it.
+	if s.next != nil {
+		s.next.ServeHTTP(rw, req)
+	}
 }
 
 // modifyRequestHeaders set or delete request headers.
-func (s *header) modifyRequestHeaders(req *http.Request) {
+func (s *Header) modifyRequestHeaders(req *http.Request) {
 	// Loop through Custom request headers
-	for header, value := range s.customRequestHeaders {
+	for header, value := range s.headers.CustomRequestHeaders {
 		if value == "" {
 			req.Header.Del(header)
 		} else {
 			req.Header.Set(header, value)
 		}
 	}
 }
+
+// ModifyResponseHeaders set or delete response headers
+func (s *Header) ModifyResponseHeaders(res *http.Response) error {
+	// Loop through Custom response headers
+	for header, value := range s.headers.CustomResponseHeaders {
+		if value == "" {
+			res.Header.Del(header)
+		} else {
+			res.Header.Set(header, value)
+		}
+	}
+
+	allowOrigin := s.getAllowOrigin()
```

Comment:
This is tied to a different instance.

https://github.com/containous/traefik/pull/3809/files#diff-8ab61f315c59ef6e9b44f0bebf69ee71R128

This is not set on this instance.